pkg/registry/registry.go

```diff
@@ -258,90 +282,79 @@ func (r *Registry) handleMirror(c *gin.Context, key string) {
 				succeeded = true
 				return nil
 			}
-			proxy.ServeHTTP(c.Writer, c.Request)
+			proxy.ServeHTTP(rw, req)
 			if !succeeded {
 				break
 			}
-			log.V(5).Info("mirrored request", "path", c.Request.URL.Path, "url", u.String())
+			log.V(5).Info("mirrored request", "url", u.String())
 			return
 		}
 	}
 }
 
-func (r *Registry) handleManifest(c *gin.Context, dgst digest.Digest) {
-	c.Set("handler", "manifest")
-	b, mediaType, err := r.ociClient.GetManifest(c, dgst)
+func (r *Registry) handleManifest(rw mux.ResponseWriter, req *http.Request, dgst digest.Digest) {
+	b, mediaType, err := r.ociClient.GetManifest(req.Context(), dgst)
 	if err != nil {
-		//nolint:errcheck // ignore
-		c.AbortWithError(http.StatusNotFound, err)
+		rw.WriteError(http.StatusNotFound, err)
 		return
 	}
-	c.Header("Content-Type", mediaType)
-	c.Header("Content-Length", strconv.FormatInt(int64(len(b)), 10))
-	c.Header("Docker-Content-Digest", dgst.String())
-	if c.Request.Method == http.MethodHead {
+	rw.Header().Set("Content-Type", mediaType)
+	rw.Header().Set("Content-Length", strconv.FormatInt(int64(len(b)), 10))
+	rw.Header().Set("Docker-Content-Digest", dgst.String())
+	if req.Method == http.MethodHead {
 		return
 	}
-	_, err = c.Writer.Write(b)
+	_, err = rw.Write(b)
 	if err != nil {
-		//nolint:errcheck // ignore
-		c.AbortWithError(http.StatusNotFound, err)
+		rw.WriteError(http.StatusNotFound, err)
 		return
 	}
 }
 
-func (r *Registry) handleBlob(c *gin.Context, dgst digest.Digest) {
-	c.Set("handler", "blob")
-	size, err := r.ociClient.Size(c, dgst)
+func (r *Registry) handleBlob(rw mux.ResponseWriter, req *http.Request, dgst digest.Digest) {
+	size, err := r.ociClient.Size(req.Context(), dgst)
 	if err != nil {
-		//nolint:errcheck // ignore
-		c.AbortWithError(http.StatusInternalServerError, err)
+		rw.WriteError(http.StatusInternalServerError, err)
 		return
 	}
-	c.Header("Content-Length", strconv.FormatInt(size, 10))
-	c.Header("Docker-Content-Digest", dgst.String())
-	if c.Request.Method == http.MethodHead {
+	rw.Header().Set("Content-Length", strconv.FormatInt(size, 10))
+	rw.Header().Set("Docker-Content-Digest", dgst.String())
+	if req.Method == http.MethodHead {
 		return
 	}
-	var w io.Writer = c.Writer
+	var w io.Writer = rw
 	if r.throttler != nil {
-		w = r.throttler.Writer(c.Writer)
```

Comment:
Throttler has the same issue of hiding ReadFrom. At the same time the throttling will have to be set on the reader side instead of writer side for this to actually work.